src/DateTimeParser.jl

```diff
@@ -0,0 +1,550 @@
+    end
+end
+function Base.tryparse(::Type{Date}, str::AbstractString; args...)
+    try
+        return Nullable{Date}(parse(Date, str; args...))
+    catch
+        return Nullable{Date}()
+    end
+end
+
+Base.parse(::Type{ZonedDateTime}, str::AbstractString; args...) = parsedate(str; args...)
+Base.parse(::Type{DateTime}, str::AbstractString; args...) = DateTime(parsedate(str; args...))
+Base.parse(::Type{Date}, str::AbstractString; args...) = Date(DateTime(parsedate(str; args...)))
+
+# Automatic parsing of DateTime strings. Based upon Python's dateutil parser
+# https://labix.org/python-dateutil#head-a23e8ae0a661d77b89dfb3476f85b26f0b30349c
+
+# Some pointers:
+# http://www.cl.cam.ac.uk/~mgk25/iso-time.html
+# http://www.w3.org/TR/NOTE-datetime
+# http://new-pds-rings-2.seti.org/tools/time_formats.html
+# http://search.cpan.org/~muir/Time-modules-2003.0211/lib/Time/ParseDate.pm
+
+const HMS = Dict{AbstractString, Symbol}(
+    "h" => :hour, "hour" => :hour, "hours" => :hour,
+    "m" => :minute, "minute" => :minute, "minutes" => :minute,
+    "s" => :second, "second" => :second, "seconds" => :second,
+)
+const AMPM = Dict{AbstractString, Symbol}(
+    "am" => :am, "a" => :am,
+    "pm" => :pm, "p" => :pm,
+)
+const JUMP = (
+    " ", ".", ",", ";", "-", "/", "'", "at", "on", "and", "ad", "m", "t", "of", "st",
+    "nd", "rd", "th", "the",
+)
+const PERTAIN = ("of",)
+const UTCZONE = ("utc", "gmt", "z",)
+
+function parsedate(datetimestring::AbstractString; fuzzy::Bool=false,
+    default::ZonedDateTime=ZonedDateTime(DateTime(year(today())), TimeZone("UTC")),
+    timezone_infos::Dict{AbstractString, TimeZone}=Dict{AbstractString, TimeZone}(), # Specify what a timezone is
+    dayfirst::Bool=false, # MM-DD-YY vs DD-MM-YY
+    yearfirst::Bool=false, # MM-DD-YY vs YY-MM-DD
+    locale::AbstractString="english", # Locale in Dates.VALUETOMONTH and VALUETODAYOFWEEK
+)
+    datetimestring = strip(datetimestring)
+
+    if isempty(datetimestring)
+        return default
+    end
+
+    res = _parsedate(datetimestring, fuzzy=fuzzy, timezone_infos=timezone_infos,
+        dayfirst=dayfirst, yearfirst=yearfirst, locale=locale)
+
+    # Fill in default values if none exits
+    res["year"] = convertyear(get(res, "year", year(default)))
+    get!(res, "month", month(default))
+    get!(res, "day", day(default))
+    get!(res, "hour", hour(default))
+    get!(res, "minute", minute(default))
+    get!(res, "second", second(default))
+    get!(res, "millisecond", millisecond(default))
+    if !haskey(res, "timezone")
+        if haskey(res, "tzoffset")
+            tzname = get(res, "tzname", "local")
+            res["timezone"] = FixedTimeZone(tzname, res["tzoffset"])
+        else
+            res["timezone"] = default.timezone
+        end
+    end
+
+    return ZonedDateTime(DateTime(res["year"], res["month"], res["day"], res["hour"],
+            res["minute"], res["second"], res["millisecond"]), res["timezone"])
+end
```

Comment:
This should probably be the `parse(::Type{ZonedDateTime}, ::AbstractString)` method. You should make 2 more of these to support DateTime and Date. I think that is preferable to always making a ZonedDateTime and then converting it.